internal/kafka/internal/services/kafka_instance_types.go

```diff
@@ -0,0 +1,84 @@
+	"github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pkg/errors"
+)
+
+//go:generate moq -out kafka_instance_types_moq.go . SupportedKafkaInstanceTypesService
+type SupportedKafkaInstanceTypesService interface {
+	GetSupportedKafkaInstanceTypesByRegion(providerId string, regionId string) ([]public.SupportedKafkaInstanceType, *errors.ServiceError)
+}
+
+type supportedKafkaInstanceTypesService struct {
+	providerConfig *config.ProviderConfig
+	kafkaConfig    *config.KafkaConfig
+}
+
+func NewSupportedKafkaInstanceTypesService(providerConfig *config.ProviderConfig, kafkaConfig *config.KafkaConfig) SupportedKafkaInstanceTypesService {
+	return &supportedKafkaInstanceTypesService{
+		providerConfig: providerConfig,
+		kafkaConfig:    kafkaConfig,
+	}
+}
+
+func (t supportedKafkaInstanceTypesService) GetSupportedKafkaInstanceTypesByRegion(providerId string, regionId string) ([]public.SupportedKafkaInstanceType, *errors.ServiceError) {
+	instanceTypeList := []public.SupportedKafkaInstanceType{}
+	provider, providerFound := t.providerConfig.ProvidersConfig.SupportedProviders.GetByName(providerId)
+	if !providerFound {
+		return nil, errors.ProviderNotSupported(fmt.Sprintf("cloud provider '%s' is unsupported", providerId))
+	}
+
+	region, regionFound := provider.Regions.GetByName(regionId)
+	if !regionFound {
+		return nil, errors.RegionNotSupported(fmt.Sprintf("cloud region '%s' is unsupported", regionId))
+	}
+
+	for k := range region.SupportedInstanceTypes {
+		instanceType, err := t.kafkaConfig.SupportedInstanceTypes.Configuration.GetKafkaInstanceTypeByID(k)
+		if err != nil {
+			return nil, errors.InstanceTypeNotSupported(fmt.Sprintf("instance type '%s' is unsupported", k))
+		}
+		sizes := instanceType.Sizes
+		supportedSizesList := []public.SupportedKafkaSize{}
+		for _, size := range sizes {
+			//errors from Quantity.ToFloat32() ignored. Strings already validated as resource.Quantity
+			ingressBytes, _ := size.IngressThroughputPerSec.ToFloat32()
+			egressBytes, _ := size.EgressThroughputPerSec.ToFloat32()
+			retentionSizeBytes, _ := size.MaxDataRetentionSize.ToFloat32()
+			supportedSizesList = append(supportedSizesList, public.SupportedKafkaSize{
+				Id: size.Id,
+				IngressThroughputPerSec: public.SupportedKafkaSizeBytesValueItem{
+					Bytes: ingressBytes,
+				},
+				EgressThroughputPerSec: public.SupportedKafkaSizeBytesValueItem{
+					Bytes: egressBytes,
+				},
+				TotalMaxConnections: int32(size.TotalMaxConnections),
+				MaxDataRetentionSize: public.SupportedKafkaSizeBytesValueItem{
+					Bytes: retentionSizeBytes,
+				},
+				MaxPartitions:               int32(size.MaxPartitions),
+				MaxDataRetentionPeriod:      size.MaxDataRetentionPeriod,
+				MaxConnectionAttemptsPerSec: int32(size.MaxConnectionAttemptsPerSec),
+				QuotaConsumed:               int32(size.QuotaConsumed),
+				QuotaType:                   size.QuotaType,
+				CapacityConsumed:            int32(size.CapacityConsumed),
+			})
```

Comment:
The service shouldn't really have to deal with any of the public api types. In case another internal component other than the handler (i.e. another service/worker) calls this function, they will get the public api type which should only be used for the API responses. Instead of having this here, there should be a presenter that handles this part, converting the internal obj to a public api obj.

Reply:
Sounds like old good DTO pattern

https://nurcahyaari.medium.com/how-i-implement-clean-code-architecture-on-golang-projects-68be58830621#:~:text=DTO%20use%20as%20a%20struct,the%20repository%2C%20and%20vice%20versa.

Reply:
Thanks @JameelB  and @wtrocki thanks for the feedback. I'll put presenter in